deepdrivemd/agents/stream/dbscan.py

```diff
@@ -0,0 +1,692 @@
+    randomly select :obj:`cfg.num_sim`, return the corresponding indices.
+
+    Parameters
+    ----------
+    cfg : OutlierDetectionConfig
+    cvae_input : Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray, np.ndarray]
+        steps, positions, velocities, md5sums, rmsds
+
+    Returns
+    -------
+    List[int]
+        List of :obj:`cfg.num_sim` indices randomly selected from a list of
+        :obj:`2*cfg.num_sim` entries with smallest rmsd.
+
+    Note
+    ----
+    This is used when no outliers are found.
+    """
+    rmsds = cvae_input[4]
+    z = sorted(zip(rmsds, range(len(rmsds))), key=lambda x: x[0])
+    sorted_index = list(map(lambda x: x[1], z))[2 * cfg.num_sim :]
+    sorted_index = random.sample(sorted_index, cfg.num_sim)
+    return sorted_index
+
+
+def select_best(
+    cfg: OutlierDetectionConfig,
+    cvae_input: Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray, np.ndarray],
+) -> List[int]:
+    """Sort cvae_input by rmsd, selects best :obj:`cfg.num_sim`, return
+    the corresponding indices.
+
+    Parameters
+    ----------
+    cfg : OutlierDetectionConfig
+    cvae_input : Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray, np.ndarray]
+        steps, positions, velocities, md5sums, rmsds
+
+    Returns
+    -------
+    List[int]
+        List of :obj:`cfg.num_sim` indices for best traversed states among
+        :obj:`lastN` from each aggregator.
+    """
+    rmsds = cvae_input[4]
+    z = sorted(zip(rmsds, range(len(rmsds))), key=lambda x: x[0])
+    sorted_index = list(map(lambda x: x[1], z))[cfg.num_sim :]
+    return sorted_index
+
+
+def main(cfg: OutlierDetectionConfig):
+    print(subprocess.getstatusoutput("hostname")[1])
+    sys.stdout.flush()
+
+    print(cfg)
+    with Timer("wait_for_input"):
+        adios_files_list = wait_for_input(cfg)
+    with Timer("wait_for_model"):
+        model_path = str(wait_for_model(cfg))
+
+    variable_list = [
+        StreamContactMapVariable("contact_map", np.uint8, 1),
+        StreamVariable("positions", np.float32, 1),
+        StreamVariable("md5", str, 2),
+        StreamVariable("velocities", np.float32, 1),
+        StreamScalarVariable("rmsd", np.float32, 0),
+    ]
+
+    mystreams = Streams(
+        adios_files_list,
+        variable_list,
+        lastN=cfg.lastN,
+        config=cfg.adios_xml_agg,
+        stream_name="AggregatorOutput",
+        batch=cfg.read_batch,
+    )
+
+    top_dir, tmp_dir, published_dir = dirs(cfg)
+    eps = cfg.init_eps
+    min_samples = cfg.init_min_samples
+
+    # Infinite loop of outlier search iterations
+    for j in itertools.count(0):
+        print(f"outlier iteration {j}")
+
+        timer("outlier_search_iteration", 1)
+
+        with Timer("outlier_read"):
+            cvae_input = mystreams.next()
+            print("=" * 10)
+            print("cvae_input")
+            print(type(cvae_input))
+            print(len(cvae_input))
+            for zzz in range(len(cvae_input)):
```

Comment:
What does zzz refer to?

Reply:
Something that was introduced for debugging/information reasons that cannot interfere with anything essential due to its absurd name. I removed those print statements.